### ISSUE TITLE
First token in timeline isn’t used as a prediction target—intended behavior?

### ISSUE THREAD

**Issue by Jiacheng98** [ipolharvard/ethos-ares#13]
When constructing `x` and `y` in `TimelineDataset.__getitem__`, the code slices the event sequence as follows:

```
python
x = torch.cat((pt_ctx, timeline[:-1]))
y = torch.cat((pt_ctx, timeline[1:]))
y[: self.context_size] = -100
```
It means that the very first token of the timeline appears in the inputs (x) but never appears in the labels (y), so the model never learns to predict that first event token from the static context. I’d like to confirm whether this is an intentional design choice or a bug. Thank you!

**Comment by prenc** [comment]
Hi @Jiacheng98,

Let me make sure first if I understand the question.

Let's assume that we have a timeline:
```
[TOK_1][TOK_2][TOK_3][TOK_3]
```
Then, we construct `x` and `y` in the following way:
```
x: [CTX_1][CTX_2][CTX_3][TOK_1][TOK_2][TOK_3]
y: [-100 ][-100 ][-100 ][TOK_2][TOK_3][TOK_4]
```
where `-100` indicates no loss.

So during training, the next token prediction looks like this:
```
[CTX_1] -> [-100 ]
[CTX_1][CTX_2] -> [-100 ]
[CTX_1][CTX_2][CTX_3] -> [-100 ]
[CTX_1][CTX_2][CTX_3][TOK_1] -> [TOK_2]
[CTX_1][CTX_2][CTX_3][TOK_1][TOK_2] -> [TOK_3]
etc.
```

So getting back to you question: yes, [TOK_1] is never predicted, because we don't want to predict first token based only on context.

Side note:
If you are interested in generation of synthetic timelines, in our most recent work, we added a new token TIMELINE_START that we use to indicated beginning of a timeline which solves this problem. Unfortunately, I haven’t uploaded a preprint yet due to lack of time. Ping me on LinkedIn if you’re interested in seeing it.

**Comment by Jiacheng98** [comment]
Thank you for the quick response @prenc ! Just to make sure I’ve got it:

1. You mentioned that the first token after the context can be arbitrary, so it isn’t predicted—is it because of the way you chunk the timeline and then append the context back onto each chunk during training?
2. In your updated setup with the new TIMELINE_START token, did you alter this process?

Thanks again!

**Comment by prenc** [comment]
@Jiacheng98 Yes, you’re right. It often happens during training that tokens from patient B are attended to by tokens from patient A. In addition, patient B’s tokens don’t have corresponding static context (e.g., demographics). This issue has been brought up several times; however, I haven’t seen a paper that shows it significantly impedes model training. We’ve achieved very high performance on downstream tasks with this setup.

One paper that builds on top of ETHOS and accounts for this issue: https://arxiv.org/abs/2505.22964
It’s definitely an interesting area to explore, but I don’t currently have the capacity to experiment with it.
